modules/conversantAnalyticsAdapter.js

```diff
@@ -0,0 +1,470 @@
+        cnvrAdUnit.sizes.push({
+          w: size[0],
+          h: size[1]
+        });
+      });
+    }
+
+    auctionEndPayload.adUnits[adUnit.code] = cnvrAdUnit;
+  });
+
+  if (Array.isArray(args.noBids)) {
+    args.noBids.forEach(noBid => {
+      const bidPayload = utils.deepAccess(auctionEndPayload, 'adUnits.' + noBid.adUnitCode + '.bids.' + noBid.bidder);
+
+      if (bidPayload !== undefined) {
+        auctionEndPayload.adUnits[noBid.adUnitCode].bids[noBid.bidder].eventCodes.push(CNVR_NO_BID);
+        auctionEndPayload.adUnits[noBid.adUnitCode].bids[noBid.bidder].timeToRespond = 0; // no info for this, would have to capture event and save it there
+      } else {
+        utils.logMessage(LOG_PREFIX + 'Unable to locate bid object via adUnitCode/bidderCode in payload for noBid reply in END_AUCTION', Object.assign({}, noBid));
+      }
+    });
+  } else {
+    utils.logError(LOG_PREFIX + 'onAuctionEnd(): noBids not defined in arguments.');
+  }
+
+  // Get bid data from bids sent
+  if (Array.isArray(args.bidsReceived)) {
+    args.bidsReceived.forEach(bid => {
+      const bidPayload = utils.deepAccess(auctionEndPayload, 'adUnits.' + bid.adUnitCode + '.bids.' + bid.bidderCode);
+      if (bidPayload !== undefined) {
+        bidPayload.eventCodes.push(CNVR_BID);
+        bidPayload.timeToRespond = bid.timeToRespond;
+        bidPayload.originalCpm = bid.originalCpm;
+        bidPayload.cpm = bid.cpm;
+        bidPayload.currency = bid.currency;
+        bidPayload.adSize = {
+          'w': bid.width,
+          'h': bid.height
+        };
+      } else {
+        utils.logMessage(LOG_PREFIX + 'Unable to locate bid object via adUnitCode/bidderCode in payload for bid reply in END_AUCTION', Object.assign({}, bid));
+      }
+    });
+  } else {
+    utils.logError(LOG_PREFIX + 'onAuctionEnd(): bidsReceived not defined in arguments.');
+  }
+
+  sendData(auctionEndPayload);
+}
+
+// =============================================== START OF HELPERS ===================================================
+
+/**
+ * Generic method to look at each key/value pair of a cache object and looks at the 'timeReceived' key, if more than
+ * the max wait time has passed then just delete the key.
+ * @param cacheObj one of our cache objects [adIdLookup or timeoutCache]
+ * @param currTime the current timestamp at the start of the most recent timer execution.
+ */
+export function cleanCache(cacheObj, currTime) {
+  Object.keys(cacheObj).forEach(key => {
+    const timeInCache = currTime - cacheObj[key].timeReceived;
+    if (timeInCache >= MAX_MILLISECONDS_IN_CACHE) {
+      delete cacheObj[key];
+    }
+  });
+}
+
+/**
+ * Helper to create an object lookup key for our timeoutCache
+ * @param auctionId id of the auction
+ * @param adUnitCode ad unit code
+ * @param bidderCode bidder code
+ * @returns string concatenation of all the params into a string key for timeoutCache
+ */
+export function getLookupKey(auctionId, adUnitCode, bidderCode) {
```

Comment:
Might not need this.  bidderRequestId appears to be unique.

Reply:
It could be, is there documentation to say that it 100% is?  It's not a UUID like auctionId or other IDs so I'm not super keen to use something that could be unique when this definitely works at the moment.

Reply:
Check out src/auction.js.  There is a typedef on BidderRequest.

```javascript
/**
 * @typedef {Object} BidderRequest
 *
 * @property {string} bidderCode - adUnit bidder
 * @property {number} auctionId - random UUID
 * @property {string} bidderRequestId - random string, unique key set on all bidRequest.bids[]
 * @property {Array.<Bid>} bids
 * @property {number} auctionStart - Date.now() at auction start
 * @property {number} timeout - callback timeout
 * @property {refererInfo} refererInfo - referer info object
 * @property {string} [tid] - random UUID (used for s2s)
 * @property {string} [src] - s2s or client (used for s2s)
 */
```

Reply:
bidderRequestId is specifically for bidders so one level beneath the bid itself.  If you look at the timeout data structure it has 'bidId' not bidderRequestId.  It looks like a similar "unique" string but if it's not a UUID there's not much guarantee it is truly unique across multiple auctions.  I've looked at several other implementations of the timeout handler and they all pretty much use a similar strategy of referencing auctionId and adUnitCode to ensure it gets mapped to the correct data.